store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt

```diff
@@ -1,157 +1,138 @@
-    private val reader: (Key) -> Flow<Output?>,
-    private val writer: suspend (Key, Input) -> Unit,
-    private val delete: (suspend (Key) -> Unit)? = null
+        private val fetcher: PipelineStore<Key, Input>,
+        private val reader: (Key) -> Flow<Output?>,
+        private val writer: suspend (Key, Input) -> Unit,
+        private val delete: (suspend (Key) -> Unit)? = null
 ) : PipelineStore<Key, Output> {
-    @ExperimentalCoroutinesApi
-    @Suppress("UNCHECKED_CAST")
-    override fun stream(request: StoreRequest<Key>): Flow<StoreResponse<Output>> {
-        return when {
-            request.shouldSkipCache(CacheType.DISK) -> fetchSkippingCache(request)
-            else -> diskNetworkCombined(request)
-        }
-    }
+    private val responseFlows = CacheBuilder
+            .newBuilder()
+            .build<StoreRequest<Key>, Pair<SendChannel<StoreResponse<Output>>, SendChannel<Command>>>()
 
     /**
-     * skipping cache, just delegate to the fetcher but update disk w/ any new data from fetcher
+     * when a new request comes in we
+     * 1. lazily create a channel for that key & make a call to disk and network
+     * 2. retain a handle to the fetcher commander allowing us to request new fetches later
+     * 3. if user wants refresh then we trigger a fetcher command
      */
-    private fun fetchSkippingCache(request: StoreRequest<Key>): Flow<StoreResponse<Output>> {
-        return fetcher.stream(request)
-            .flatMapLatest { response: StoreResponse<Input> ->
-                // explicit type is necessary for type resolution
-                flow<StoreResponse<Output>> {
-                    val data = response.dataOrNull()
-                    if (data != null) {
-                        // save into database first
-                        writer(request.key, response.requireData())
-                        // continue database data
-                        var first = true
-                        val readerFlow: Flow<StoreResponse<Output>> =
-                            reader(request.key).mapNotNull {
-                                it?.let {
-                                    val origin = if (first) {
-                                        first = false
-                                        response.origin
-                                    } else {
-                                        ResponseOrigin.Persister
-                                    }
-                                    StoreResponse.Data(
-                                        value = it,
-                                        origin = origin
-                                    )
+    @ExperimentalCoroutinesApi
+    @Suppress("UNCHECKED_CAST")
+    override fun stream(request: StoreRequest<Key>): Flow<StoreResponse<Output>> = channelFlow {
+        val response = responseFlows
+                .get(request) {
+                    val sink  = actor<StoreResponse<Output>>(capacity = Channel.UNLIMITED) {
```

Comment:
taking this actor out of the channel flow might be a good idea to cleanup the code. I'm not sure where send goes here (i guess same problem exists in current pipeline persister)

Reply:
the problem is that actor needs a scope so right now we are using the channelFlow scope